pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

```diff
@@ -1841,242 +1845,302 @@ func getNumberOfPodInterfaces(spec *v1.VirtualMachineInstanceSpec) int {
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("disk %s has malformed PCI address (%s).", field.Child("domain", "devices", "disks", "disk").Index(idx).Child("name").String(), disk.Disk.PciAddress),
+			Field:   field.Child("domain", "devices", "disks", "disk").Index(idx).Child("pciAddress").String(),
+		})
+	}
+	return causes
+}
 
-				// special case. virtio is incompatible with CD-ROM for q35 machine types
-				if diskType == "cdrom" && bus == v1.DiskBusVirtio {
-					causes = append(causes, metav1.StatusCause{
-						Type:    metav1.CauseTypeFieldValueInvalid,
-						Message: fmt.Sprintf("Bus type %s is invalid for CD-ROM device", bus),
-						Field:   field.Index(idx).Child("cdrom", "bus").String(),
-					})
+func validateBootOrderValue(field *k8sfield.Path, idx int, disk v1.Disk) []metav1.StatusCause {
+	var causes []metav1.StatusCause
+	if disk.BootOrder != nil && *disk.BootOrder < 1 {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("%s must have a boot order > 0, if supplied", field.Index(idx).String()),
+			Field:   field.Index(idx).Child("bootOrder").String(),
+		})
+	}
+	return causes
+}
 
-				}
-				// sata disks (in contrast to sata cdroms) don't support readOnly
-				if disk.Disk != nil && bus == v1.DiskBusSATA && disk.Disk.ReadOnly {
-					causes = append(causes, metav1.StatusCause{
-						Type:    metav1.CauseTypeFieldValueInvalid,
-						Message: fmt.Sprintf("%s hard-disks do not support read-only.", bus),
-						Field:   field.Index(idx).Child("disk", "bus").String(),
-					})
-				}
-			}
+func getDiskBus(disk v1.Disk) string {
+	switch {
+	case disk.Disk != nil:
+		return string(disk.Disk.Bus)
+	case disk.LUN != nil:
+		return string(disk.LUN.Bus)
+	case disk.CDRom != nil:
+		return string(disk.CDRom.Bus)
+	default:
+		return ""
+	}
+}
 
-			// Reject defining DedicatedIOThread to a disk without VirtIO bus since this configuration
-			// is not supported in libvirt.
-			isIOThreadsWithUnsupportedBus := disk.DedicatedIOThread != nil && *disk.DedicatedIOThread &&
-				bus != v1.DiskBusVirtio
-			if isIOThreadsWithUnsupportedBus {
-				causes = append(causes, metav1.StatusCause{
-					Type:    metav1.CauseTypeFieldValueNotSupported,
-					Message: fmt.Sprintf("IOThreads are not supported for disks on a %s bus", bus),
-					Field:   field.Child("domain", "devices", "disks").Index(idx).String(),
-				})
-			}
-		}
+func getDiskType(disk v1.Disk) string {
+	switch {
+	case disk.Disk != nil:
+		return "disk"
+	case disk.LUN != nil:
+		return "lun"
+	case disk.CDRom != nil:
+		return "cdrom"
+	default:
+		return ""
+	}
+}
 
-		// Verify serial number is made up of valid characters for libvirt, if provided
-		if disk.Serial != "" && !isValidExpression(disk.Serial) {
+func validateBusSupport(field *k8sfield.Path, idx int, disk v1.Disk) []metav1.StatusCause {
+	var causes []metav1.StatusCause
+	bus := v1.DiskBus(getDiskBus(disk))
```

Comment:
If we don't do it then we will need to implement the switch statements in the following way
```
bus := getDiskBus(disk)
switch bus {
case string(v1.DiskBusVirtio):
    ...
case string(v1.DiskBusSATA):
    ...
...
}
```
Instead of casting in each case statement, felt that it would be better to cast here itself in only one place.

Reply:
IMHO does keeping the return type of `getDiskBus` as `v1.DiskBus` instead of `string` make things better? It will avoid any need to inter casting between `string` and `v1.DiskBus`. And keep the return type for `getDiskBus` consistent with its name. The function will look like the following if we make this change
```
func getDiskBus(disk v1.Disk) v1.DiskBus {
	switch {
	case disk.Disk != nil:
		return disk.Disk.Bus
	case disk.LUN != nil:
		return disk.LUN.Bus
	case disk.CDRom != nil:
		return disk.CDRom.Bus
	default:
		return ""
	}
}
```

Reply:
Agree with you, I'd keep `getDiskBus` as is.